Mathlib/RingTheory/Localization/Finiteness.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.Algebra.Module.LocalizedModuleIntegers
+import Mathlib.RingTheory.LocalProperties
+
+/-!
+
+# Finiteness properties under localization
+
+In this file we establish behaviour of `Module.Finite` under localizations.
+
+## Main results
+
+- `Module.Finite.of_isLocalizedModule`: If `M` is a finite `R`-module,
+  `S` is a submonoid of `R`, `Rₚ` is the localization of `R` at `S`
+  and `Mₚ` is the localization of `M` at `S`, then `Mₚ` is a finite
+  `Rₚ`-module.
+- `Module.Finite.of_localizationSpan_finite`: If `M` is an `R`-module
+  and `{ r }` is a finite set generating the unit ideal such that
+  `Mᵣ` is a finite `Rᵣ`-module for each `r`, then `M` is a finite `R`-module.
+
+-/
+
+universe u v w t
+
+open Classical
+
+namespace Module.Finite
+
+section
+
+variable {R : Type u} [CommSemiring R] (S : Submonoid R)
+variable {Rₚ : Type v} [CommSemiring Rₚ] [Algebra R Rₚ] [IsLocalization S Rₚ]
+variable {M : Type w} [AddCommMonoid M] [Module R M]
+variable {Mₚ : Type t} [AddCommMonoid Mₚ] [Module R Mₚ] [Module Rₚ Mₚ] [IsScalarTower R Rₚ Mₚ]
+variable (f : M →ₗ[R] Mₚ) [IsLocalizedModule S f]
+
+lemma of_isLocalizedModule [Module.Finite R M] : Module.Finite Rₚ Mₚ := by
+  obtain ⟨T, hT⟩ := ‹Module.Finite R M›
+  use T.image f
+  rw [eq_top_iff]
+  rintro x -
+  obtain ⟨⟨y, m⟩, (hyx : IsLocalizedModule.mk' f y m = x)⟩ :=
+    IsLocalizedModule.mk'_surjective S f x
+  have hy : y ∈ Submodule.span R T := by rw [hT]; trivial
+  have : f y ∈ Submodule.map f (Submodule.span R T) := Submodule.mem_map_of_mem hy
+  rw [Submodule.map_span] at this
+  have H : Submodule.span R (f '' T) ≤
+      (Submodule.span Rₚ (f '' T)).restrictScalars R := by
+    rw [Submodule.span_le]; exact Submodule.subset_span
+  convert (Submodule.span Rₚ (f '' T)).smul_mem
+    (IsLocalization.mk' Rₚ (1 : R) m) (H this) using 1
+  · rw [← hyx, ← IsLocalizedModule.mk'_one S, IsLocalizedModule.mk'_smul_mk']
+    simp
+  · simp
+
+end
+
+variable {R : Type u} [CommRing R] (S : Submonoid R) {M : Type w} [AddCommMonoid M] [Module R M]
+
+/--
+If there exists a finite set `{ r }` of `R` such that `Mᵣ` is `Rᵣ`-finite for each `r`,
+then `M` is a finite `R`-module.
+
+General version for any modules `Mᵣ` and rings `Rᵣ` satisfying the correct universal properties.
+See `Module.Finite.of_localizationSpan_finite` for the specialized version.
+-/
+theorem of_localizationSpan_finite' (t : Finset R) (ht : Ideal.span (t : Set R) = ⊤)
```

Comment:
Can you give a `Set R` version? (using `Ideal.span_eq_top_iff_finite`)

Reply:
Done.